zkevm-circuits/src/copy_circuit.rs

```diff
@@ -0,0 +1,625 @@
+                - meta.query_advice(is_last, Rotation::cur())
+                - meta.query_advice(is_last, Rotation::next());
+            cb.condition(not_last_two_rows, |cb| {
+                cb.require_equal(
+                    "rows[0].id == rows[2].id",
+                    meta.query_advice(id, Rotation::cur()),
+                    meta.query_advice(id, Rotation(2)),
+                );
+                cb.require_equal(
+                    "rows[0].tag == rows[2].tag",
+                    tag.value(Rotation::cur())(meta),
+                    tag.value(Rotation(2))(meta),
+                );
+                cb.require_equal(
+                    "rows[0].addr + 1 == rows[2].addr",
+                    meta.query_advice(addr, Rotation::cur()) + 1.expr(),
+                    meta.query_advice(addr, Rotation(2)),
+                );
+            });
+
+            let rw_diff = and::expr([
+                or::expr([
+                    tag.value_equals(CopyDataType::Memory, Rotation::cur())(meta),
+                    tag.value_equals(CopyDataType::TxLog, Rotation::cur())(meta),
+                ]),
+                not::expr(meta.query_advice(is_pad, Rotation::cur())),
+            ]);
+            cb.condition(
+                not::expr(meta.query_advice(is_last, Rotation::cur())),
+                |cb| {
+                    cb.require_equal(
+                        "rows[0].rw_counter + rw_diff == rows[1].rw_counter",
+                        meta.query_advice(rw_counter, Rotation::cur()) + rw_diff.clone(),
+                        meta.query_advice(rw_counter, Rotation::next()),
+                    );
+                    cb.require_equal(
+                        "rows[0].rwc_inc_left - rw_diff == rows[1].rwc_inc_left",
+                        meta.query_advice(rwc_inc_left, Rotation::cur()) - rw_diff.clone(),
+                        meta.query_advice(rwc_inc_left, Rotation::next()),
+                    );
+                },
+            );
+            cb.condition(meta.query_advice(is_last, Rotation::cur()), |cb| {
+                cb.require_equal(
+                    "rwc_inc_left == rw_diff for last row in the copy slot",
+                    meta.query_advice(rwc_inc_left, Rotation::cur()),
+                    rw_diff,
+                );
+            });
+
+            cb.gate(meta.query_fixed(q_enable, Rotation::cur()))
+        });
+
+        meta.create_gate("verify step", |meta| {
+            let mut cb = BaseConstraintBuilder::default();
+
+            cb.require_zero(
+                "bytes_left == 1 for last step",
+                and::expr([
+                    meta.query_advice(is_last, Rotation::next()),
+                    1.expr() - meta.query_advice(bytes_left, Rotation::cur()),
+                ]),
+            );
+
+            cb.condition(
+                not::expr(meta.query_advice(is_last, Rotation::next())),
+                |cb| {
+                    cb.require_equal(
```

Comment:
This constraint is under the condition `not(is_last_two_rows)` and without `q_step` in the spec: https://github.com/privacy-scaling-explorations/zkevm-specs/blob/5426f6a77b42c51b294ea5988e4a23e4ce0ec0e6/src/zkevm_specs/copy_circuit.py#L44

Probably this is correct, because `src_addr_end` is only used on `q_step = 1`, but I just wanted to check :)

Reply:
Ah! I am not sure how this constraint ended up in the wrong condition.

But it still worked correctly because `q_step == true && is_last == false` only ignores the last 2 rows. Fixed it in the last commit https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/584/commits/6ff3dbe38b2b395ee409936ec7e26c87ca242252